historical_system_profiles/archiver.py

```diff
@@ -84,6 +92,68 @@ def _archive_profile(data, ptc, logger, notification_service):
             "wrote %s to historical database (inv id: %s, captured_on: %s)"
             % (hsp.id, hsp.inventory_id, hsp.captured_on)
         )
+        # After the new hsp is saved, we need to check for any reason to alert via
+        # triggering a notification, i.e. if drift from any associated baselines has
+        # occurred.
+        _check_and_send_notifications(
+            host["id"],
+            host["account"],
+            host["insights_id"],
+            host["updated"],
+            host["display_name"],
+            host["tags"],
+            service_auth_key,
+            logger,
+        )
+
+
+def _check_and_send_notifications(
+    inventory_id,
+    account_id,
+    insights_id,
+    system_check_in,
+    display_name,
+    tags,
+    service_auth_key,
+    logger,
+):
+    # After the new hsp is saved, we need to check for any reason to alert Notifications
+    # First, check if this system id is associated with any baselines
+    # If yes, then for each baseline associated, call for a short-circuited comparison
+    # to see if the system has drifted from that baseline.
+    # If anything has changed, send a kafka message to trigger a notification.
+
+    baseline_ids = fetch_system_baseline_associations(
+        inventory_id, service_auth_key, logger
+    )
+    # If yes, then for each baseline associated, call for a short-circuited comparison
+    # to see if the system has drifted from that baseline.
+    if baseline_ids:
+        drift_found = False
+        event = EventDriftBaselineDetected(
+            account_id, insights_id, system_check_in, display_name, tags
+        )
+        for baseline_id in baseline_ids:
+            if check_for_drift(inventory_id, baseline_id, service_auth_key, logger):
```

Comment:
check_for_drift() returns whatever fetch_url returns.  This if statement  should always evaluate to True.  I think you need to change the name from check_for_drift() to get_system_baseline_comparison_report() and assign the return to a variable.  Then iterate over the data returned until you find a fact that is in a state of "DIFFERENT".  If you find that state, this system-baseline combination needs to be added to the event message.

Reply:
The problem is that we currently mark state different if the fact is not on the baseline, so we need to parse down to only the facts on this baseline, and the short_circuit took care of handling that here:
https://github.com/RedHatInsights/drift-backend/pull/238/files#diff-d94d1e0636fff7ab5a4d4f6f24ff091770a8a2ec2c170453ec644766a4f16857R207-R213
by only sending in the baseline facts.
Since we want to send comparison report info along as part of the payload in the future, I think this could be rewritten to then have to go through every fact on the baseline and cross reference to the comparison report in here, or we can do a rewrite of our comparison report logic and call a different method or something, but for now for the sake of time, I think we opted to move forward with getting this to just start with the baseline_id.
That being said, you're absolutely right that this is "truthy" all the time and that needs to be adjusted; it had been on the to do list but forgotten, so with the schema fix on drift-backend, I'll revisit what we return, maybe add a boolean key to the report that can then be checked for here.

Reply:
Actually, looking at it again, it's not always truthy, we were returning False if there was drift, but it's a moot point as to make the schema happy I think we're going to have to add a key in the report instead of just returning a bool in place of the report.  We definitely need to have a larger discussion on the entire comparison report's logic during cleanup.

Reply:
Updated to return the comparison report with drift_event_notify as a new dict key.  How's this look?

Reply:
Related drift changes:
https://github.com/RedHatInsights/drift-backend/pull/248/

Reply:
This should work.  Main thing was that fetch_url had always returned something (in json) that would result in the if always being true.